In [1]:
from scipy import stats
import numpy as np

In [2]:
dir(stats)

['F_onewayBadInputSizesWarning',
 'F_onewayConstantInputWarning',
 'NumericalInverseHermite',
 'PearsonRConstantInputWarning',
 'PearsonRNearConstantInputWarning',
 'SpearmanRConstantInputWarning',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '_binned_statistic',
 '_binomtest',
 '_boost',
 '_bootstrap',
 '_common',
 '_constants',
 '_continuous_distns',
 '_crosstab',
 '_discrete_distns',
 '_distn_infrastructure',
 '_distr_params',
 '_entropy',
 '_hypotests',
 '_ksstats',
 '_mannwhitneyu',
 '_multivariate',
 '_page_trend_test',
 '_qmc',
 '_qmc_cy',
 '_relative_risk',
 '_rvs_sampling',
 '_sobol',
 '_stats',
 '_stats_mstats_common',
 '_tukeylambda_stats',
 '_wilcoxon_data',
 'alexandergovern',
 'alpha',
 'anderson',
 'anderson_ksamp',
 'anglit',
 'ansari',
 'arcsine',
 'argus',
 'barnard_exact',
 'bartlett',
 'bayes_mvs',
 'bernoulli',
 'beta',
 'betabinom',
 'betaprime',
 'biasedurn',
 'binned_s

In [3]:
help(stats)

Help on package scipy.stats in scipy:

NAME
    scipy.stats - .. _statsrefmanual:

DESCRIPTION
    Statistical functions (:mod:`scipy.stats`)
    
    .. currentmodule:: scipy.stats
    
    This module contains a large number of probability distributions,
    summary and frequency statistics, correlation functions and statistical
    tests, masked statistics, kernel density estimation, quasi-Monte Carlo
    functionality, and more.
    
    Statistics is a very large area, and there are topics that are out of scope
    for SciPy and are covered by other packages. Some of the most important ones
    are:
    
    - `statsmodels <https://www.statsmodels.org/stable/index.html>`__:
      regression, linear models, time series analysis, extensions to topics
      also covered by ``scipy.stats``.
    - `Pandas <https://pandas.pydata.org/>`__: tabular data, time series
      functionality, interfaces to other statistical languages.
    - `PyMC3 <https://docs.pymc.io/>`__: Bayesian statistica

In [60]:
import pandas as pd
from scipy.stats import ttest_1samp, ttest_ind, ttest_rel, shapiro, wilcoxon, f_oneway, chi2_contingency, anderson, levene, bartlett

 다음 데이터의 정규성을 검증해라

In [5]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/normal1.csv')
df

data
0    295.979767
1    289.703739
2    300.609670
3    299.890856
4    293.159612
..          ...
995  299.901120
996  307.387808
997  303.869390
998  313.979589
999  306.950007

[1000 rows x 1 columns]

In [6]:
shapiro(df['data'])

ShapiroResult(statistic=0.9981446862220764, pvalue=0.3485966622829437)

-> 정규성을 가진다 (귀무가설) p>0.05 ->귀무가설 채택

-> 정규성을 가지지 않는다 (대립가설)

In [7]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/normal3.csv')
df

data
0    0.880145
1   -0.021619
2    0.509888
3    0.179442
4    0.493778
..        ...
995 -0.229930
996  0.087762
997  0.651349
998  2.489877
999  0.704755

[1000 rows x 1 columns]

In [8]:
s, p = shapiro(df['data'])
p

2.309859641255865e-16

In [9]:
if p > 0.05 :
    print('귀무가설 채택')
else : 
    print('대립가설 채택 (정규성을 가지지 않는다)')

대립가설 채택 (정규성을 가지지 않는다)


위의 데이터를 log변환 한 후에 정규성을 가지는지 확인하라

In [10]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/normal3.csv')
df

data
0    0.880145
1   -0.021619
2    0.509888
3    0.179442
4    0.493778
..        ...
995 -0.229930
996  0.087762
997  0.651349
998  2.489877
999  0.704755

[1000 rows x 1 columns]

In [13]:
shapiro(np.log1p(df))

ShapiroResult(statistic=0.9976893067359924, pvalue=0.17551212012767792)

In [15]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/normal6.csv')
df

data
0     296.016803
1     299.628476
2     298.232057
3     293.256749
4     301.037710
...          ...
5995  304.008424
5996  301.043219
5997  302.542420
5998  294.456145
5999  301.733819

[6000 rows x 1 columns]

In [21]:
anderson(df['data'])

AndersonResult(statistic=0.8266993530414766, critical_values=array([0.576, 0.656, 0.786, 0.917, 1.091]), significance_level=array([15. , 10. ,  5. ,  2.5,  1. ]))

유의확률 0.05 인 경우 significance_level 에서 5. 인 부분에서의 critival_values 찾기 -> 여기서는 0.786

---   > 0.82 -> 귀무가설 채택 

0.786 < 0.82 -> 대립가설 채택

-----------------------------------------------------------

100명의 키 정보가 들어 있는 데이터가 있다.데이터가 정규성을 만족하는지 확인하라.

그리고 평균키는 165라 판단할수 있는지 귀무가설과 대립가설을 설정한 후 유의수준 5%로 검정하라.

In [23]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/height1.csv')
df

height
0   160.237691
1   164.747324
2   165.401628
3   168.801627
4   153.199021
..         ...
95  168.556633
96  172.935092
97  171.913052
98  181.381108
99  161.891018

[100 rows x 1 columns]

In [24]:
shapiro(df)

ShapiroResult(statistic=0.9872670769691467, pvalue=0.455832302570343)

In [26]:
ttest_1samp(df['height'], popmean = 165) 
# -> 대립가설 채택 (165라 판단할 수 없음)

Ttest_1sampResult(statistic=3.2017884987150644, pvalue=0.0018367171548080209)

In [27]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/height2.csv')
df

height
0   182.012475
1   166.857081
2   159.633265
3   159.633265
4   169.428478
..         ...
95  160.237691
96  159.633265
97  174.777363
98  159.633265
99  164.747324

[100 rows x 1 columns]

In [35]:
shapiro(df)
# p<0.05 정규성을 만족하지 않음

ShapiroResult(statistic=0.9672006368637085, pvalue=0.013552471995353699)

In [36]:
s, p = wilcoxon(df['height'])
p

3.87726171873134e-18

In [37]:
p <0.05
# 평균은 165라 판단할 수 있음

True

두개 학급의 시험성적에 대한 데이터이다 그룹간 등분산 검정을 시행하라 nan값 삭제해야 함 df.B.dropna()

In [38]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/scipy2.csv')
df.head()

class  score
0     A     84
1     A     59
2     A     49
3     A     57
4     A     82

In [46]:
df['class'].value_counts()

A    300
B    180
Name: class, dtype: int64

In [51]:
adf = df[df['class'] == 'A'].score
bdf = df[df['class'] == 'B'].score

In [52]:
levene(adf, bdf)

LeveneResult(statistic=0.3145466542912649, pvalue=0.5751662820554713)

In [53]:
# p>0.05 등분산성 만족함

두개 학급의 시험성적에 대한 데이터이다 그룹간 등분산 검정을 시행하라

In [54]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/scipy3.csv')
df.head()

class  score
0     A     32
1     A     69
2     A     68
3     A     60
4     A     78

In [55]:
df['class'].value_counts()

A    300
B    180
Name: class, dtype: int64

In [56]:
adf = df[df['class'] == 'A'].score
bdf = df[df['class'] == 'B'].score

In [57]:
levene(adf, bdf)

LeveneResult(statistic=4.307122424591436, pvalue=0.03848734007752694)

In [58]:
# p<0.05 등분산성을 만족하지 않음 -> 대립가설 채택

In [59]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/scipy6.csv')
df.head()

A   B
0  44  36
1  61  92
2  76  41
3  64  45
4  65  35

In [62]:
bartlett(df['A'], df['B'])
# p > 0.05 등분산성이다

BartlettResult(statistic=2.3832178811043527, pvalue=0.12264468401745829)

## 독립표본t검정

두개 학급의 시험성적에 대한 데이터이다. 두 학습의 시험 평균(비모수검정의 경우 중위값)은 동일하다 말할 수 있는지 확인 하라

In [65]:
df1 = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/ind1.csv')
df2 = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/ind2.csv')

In [70]:
print(shapiro(df1)) #-> 정규성 만족
print(shapiro(df2)) #-> 정규성 만족

ShapiroResult(statistic=0.9860947728157043, pvalue=0.37968069314956665)
ShapiroResult(statistic=0.990182638168335, pvalue=0.6793646216392517)


In [74]:
levene(df1['data'], df2['data']) # 등분산성 만족

LeveneResult(statistic=2.5337683795339547, pvalue=0.11302904824469093)

In [78]:
ttest_ind(df1, df2, equal_var=True) #띄어쓰기 안 함

Ttest_indResult(statistic=array([2.76719074]), pvalue=array([0.00619015]))

In [79]:
# p<0.05 -> 대립가설 채택 -> 시험 평균이 동일하다고 말할 수 없음

두개 학급의 시험성적에 대한 데이터이다. 두 학습의 시험 평균(비모수검정의 경우 중위값)은 동일하다 말할 수 있는지 확인 하라

In [97]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/scipy5.csv')
df

A     B
0    74  30.0
1    30  59.0
2    35  69.0
3    45  88.0
4    74  68.0
..   ..   ...
295  80   NaN
296  30   NaN
297  91   NaN
298  84   NaN
299  57   NaN

[300 rows x 2 columns]

In [102]:
s, p = shapiro(df.A.dropna())
p

6.175894240456614e-10

In [103]:
p<0.05 # 정규성을 만족하지 않는다 -> 독립표본 t-검정을 할 수 없음

True

In [104]:
s, p = shapiro(df.B.dropna())
p

0.00013572089665103704

In [105]:
p<0.05 # 정규성을 만족하지 않는다 -> 독립표본 t-검정을 할 수 없음

True

In [106]:
from scipy.stats import mannwhitneyu , ranksums
print()
print(mannwhitneyu(df['A'],df['B'].dropna()))
print(ranksums(df['A'],df['B'].dropna()))


MannwhitneyuResult(statistic=27036.0, pvalue=0.9807458376150018)
RanksumsResult(statistic=0.02446942170858557, pvalue=0.9804781743503561)


In [107]:
# p>0.05 -> 귀무가설 채택 ->평균이 동일하다고 말할 수 있다

두개 그룹에 대한 수치형 데이터이다. 두 그룹의 평균은 동일하다 말할 수 있는지 검정하라

In [108]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/ind3.csv')
df

group        data
0       a  108.065847
1       a   88.714106
2       a   61.182748
3       a   51.763444
4       a  127.858730
..    ...         ...
185     b  100.192985
186     b  112.081790
187     b   83.148014
188     b  147.504761
189     b  117.412623

[190 rows x 2 columns]

In [109]:
adf = df[df['group'] == 'a'].data
bdf = df[df['group'] == 'b'].data

In [110]:
shapiro(adf)

ShapiroResult(statistic=0.9834123253822327, pvalue=0.1473984718322754)

In [111]:
shapiro(bdf)

ShapiroResult(statistic=0.9831849336624146, pvalue=0.47018250823020935)

In [112]:
# 둘 다 정규성 만족 -> 이제 등분산성 확인하기

In [113]:
levene(adf, bdf)

LeveneResult(statistic=6.185601018015722, pvalue=0.013750484571911342)

In [114]:
# 등분산성 만족하지 않음 -> equal_val=False

In [117]:
ttest_ind(adf, bdf, equal_var=False)

Ttest_indResult(statistic=-2.1949470315829265, pvalue=0.029512802991767915)

In [118]:
# p<0.05 -> 대립가설 채택 (평균이 동일하다고 말할 수 없음)

In [119]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/ind6.csv')
df

a           b
0    121.023118   69.982090
1     69.740978  139.774624
2     93.737693  175.425924
3    100.331645   85.493555
4     75.048925   83.488953
..          ...         ...
115   97.955501         NaN
116  116.839209         NaN
117   88.257448         NaN
118   89.904714         NaN
119   88.250342         NaN

[120 rows x 2 columns]

In [120]:
a = df['a'].dropna()
b = df['b'].dropna()

In [121]:
print(shapiro(a))
print(shapiro(b))

ShapiroResult(statistic=0.9865895509719849, pvalue=0.28390026092529297)
ShapiroResult(statistic=0.9854329228401184, pvalue=0.5937777757644653)


In [123]:
levene(a, b)

LeveneResult(statistic=3.9862856894158347, pvalue=0.04731495612868527)

In [124]:
ttest_ind(a, b, equal_var=False)

Ttest_indResult(statistic=0.0015963310698567188, pvalue=0.9987289046092704)

In [125]:
# 평균 동일하다고 말할 수 있다(귀무가설 채택)

## 대응표본 t 검정 (paired)

특정 질병 집단의 투약 전후의 혈류량 변화를 나타낸 데이터이다. 투약 전후의 변화가 있는지 검정하라

In [127]:
df = pd.read_csv('https://raw.githubusercontent.com/Datamanim/datarepo/main/scipy/rel2.csv')
df

before       after
0    216.289726  175.897620
1    190.279454  189.295934
2    217.744689  219.233935
3    235.021264  195.021021
4    234.098484  213.218227
..          ...         ...
115  186.546583  239.845621
116  199.853014  300.032396
117  182.322187  201.310529
118  256.739402  195.169040
119  248.858766  292.766655

[120 rows x 2 columns]

In [128]:
before = df['before']
after = df['after']

In [132]:
print(shapiro(before), shapiro(after))

ShapiroResult(statistic=0.9907894134521484, pvalue=0.6065712571144104) ShapiroResult(statistic=0.9916958808898926, pvalue=0.6923439502716064)


In [134]:
print(levene(before, after))

LeveneResult(statistic=0.06427968690211128, pvalue=0.8000741651677987)


In [137]:
ttest_rel(before, after, alternative='less')

Ttest_relResult(statistic=-2.5535473487670677, pvalue=0.005963372362273256)

In [131]:
# 유의미하다 (대립가설 채택, 약이 효과가 있었다!!)

특정 집단의 학습 전후 시험 성적 변화를 나타낸 데이터이다. 시험 전과 후에 차이가 있는지 검정하라